File: 2025/01/secret.py
""" Advent of Code 2025. Day 1: Secret Entrance """


with open("input.txt") as f:
    rotations = f.read().splitlines()
    # convert to negative/positive steps
    rotations = [+int(r[1:]) if r[0] == "R" else -int(r[1:]) for r in rotations]



def find_password1(rotations):
    dial = 50

    password = 0

    for r in rotations:
        dial += r
        dial %= 100

        if dial == 0:
            password += 1
    return password

print("Part 1:\t", find_password1(rotations))

def find_password2(rotations):
    dial = 50
    password = 0

    for r in rotations:
        while abs(r) > 100:
            password += 1

            if r > 0:
                r -= 100
            else:
                r += 100

        prev_dial = dial
        dial += r
        dialmod = dial % 100

        if dialmod == 0:
            password += 1
        elif dialmod != dial and prev_dial != 0:
            password += 1
        dial = dialmod
    return password

print("Part 2:\t", find_password2(rotations))

File: 2025/01/secret.f90
program day01
    implicit none

    ! Variables related to file I/O
    integer :: iostat
    integer :: io
    integer :: length = 0

    ! Problem related variables
    character*256, dimension(:), allocatable :: scanning
    integer :: dial = 50
    integer :: password = 0


    ! Fortran needs to know how large the array is going to be
    ! before allocating
    ! First we do a pass over the input file to count number of lines
    open(newunit=io, file="input.txt", status="old", action="read")
    do
        read(io, *, iostat=iostat) 
        if (is_iostat_end(iostat)) exit
        length = length + 1
    end do
    ! Now we can allocate our scanning array
    allocate(scanning(length))
    ! And finally, we can store the input in the array
    rewind(io)
    read(io, *) scanning
    close(io)


    call part1(scanning)

    write (*,*) "Part 1", password

  contains

    impure elemental subroutine part1(rotation)
      implicit none

      character*256, intent(in) :: rotation

      character :: dir
      integer :: steps

      dir = rotation(1:1)
      ! convert the rest of the string to an integer
      read(rotation(2:), *) steps

      ! This block can be replaced by merge(),
      ! but the if block is more readable
      if (dir == "L") then
        dial = dial - steps
      else
        dial = dial + steps
      end if
      dial = modulo(dial, 100)

      if (dial == 0) then
        password = password + 1
      end if
      
    end subroutine part1
end program day01
